### DISCUSSION TITLE
Using PKCS#11 with OpenSSL

### DISCUSSION THREAD

**Discussion by Hussain1811** [openssl/openssl#24472] [Q&A] [answered]
Do we have an engine or a provider for the same?

**Comment by levitte** [comment, marked as answer]
@latchset make a provider, here: https://github.com/latchset/pkcs11-provider

**Reply by Hussain1811** [reply]
Is it preferable to use the Engine (`engine_pkcs11` via https://github.com/OpenSC/libp11) or the provider ( https://github.com/latchset/pkcs11-provider) ?

**Reply by levitte** [reply]
That depends on the OpenSSL version, I guess.  For OpenSSL 3.0 and newer, providers are the path forward

**Reply by paulidale** [reply]
Create a provider.  No supported version of OpenSSL recommends engines.  They are deprecated in 3.0 and will be removed in due course.

**Reply by Hussain1811** [reply]
Can we configure OpenSSL to use this provider via `openssl.cnf`?

**Reply by beldmit** [reply]
Yes

**Comment by Hussain1811** [comment]
i was able to configure the PKCS11 provider of latchset. `openssl list -providers` lists it too.
Now, how do i play around with this provider via the openssl cmd-line tool for PKCS11 operations? (listing slots, tokens, generating keys, doing crypto operations etc)

**Reply by Hussain1811** [reply]
Please correct me if i'm wrong. I guess we need to use the `PKCS#11 URI scheme` (https://www.rfc-editor.org/rfc/rfc7512) in order to use PKCS#11 with OpenSSL. 
So, i pointed `pkcs11-module-path` to `libopencryptoki.so` and ended up in below error:
```
/opt/openssl/bin/openssl pkey -in "pkcs11:type=public" -pubin -pubout -out testkey.pub -provider pkcs11
Could not find private key of Public Key from pkcs11:type=public
```

**Reply by levitte** [reply]
If you try `/opt/openssl/bin/openssl storeutl "pkcs11:type=public"`, what do you get?

Generally speaking, `openssl storeutl` is the command to inspect and discover URI contents, do try it out!

**Reply by Hussain1811** [reply]
```
/opt/openssl/bin/openssl storeutl "pkcs11:type=public"
Total found: 0
```

**Reply by levitte** [reply]
Okie, so the provider lists nothing for that URI.  This *may* be a lack in the PKCS#11 provider, it *may* also be that it's not very workable to list very much from that URI.
I'm not sure if PKCS#11 allows this, but if it does, `openssl storeutl -r "pkcs11:type=public"` would recurse into "deeper" URIs, i.e. figure out something that could be used as a directory tree, but that would require that "pkcs11:type=public" itself returns *at the very least* one "deeper" URI for it to follow...  This could be used to get a list of all entries, including slots and stuff, at the discretion of the provider.

Thing is that `openssl pkey` expects the `-in` parameter to be a file or URI that contains at least one private key, without recursion.  As you can see for yourself, that's not the case.

**Reply by Hussain1811** [reply]
Thanks for the information @levitte 
Any idea how i can proceed further? 
i even used a different PKCS#11 module (other than libopencryptoki.so) - i used the `NSS softokn`. But, the results are the same.

**Reply by t8m** [reply]
p11tool from GnuTLS should be able to list objects on a token with URIs.
